kotlinx-coroutines-core/common/src/channels/Channels.common.kt

```diff
@@ -189,11 +200,68 @@ public suspend inline fun <E> ReceiveChannel<E>.consumeEach(action: (E) -> Unit)
  * ```
  */
 public suspend fun <E> ReceiveChannel<E>.toList(): List<E> = buildList {
-    consumeEach {
-        add(it)
-    }
+    consumeEach(::add)
 }
 
+/**
+ * [Consumes][consume] the elements of this channel into the provided mutable collection.
+ *
+ * This is a convenience function equivalent to calling [consumeAsFlow]
+ * followed by [kotlinx.coroutines.flow.toCollection].
+ * Please use [consumeAsFlow] directly in scenarios where elements should undergo additional transformations
+ * before being added to the resulting collection.
+ *
+ * [consumeTo] attempts to receive elements and put them into the collection until the channel is
+ * [closed][SendChannel.close].
+ *
+ * If the channel is [closed][SendChannel.close] with a cause, [consumeTo] will rethrow that cause.
+ * However, the elements already received up to that point will remain in the collection.
+ *
+ * Since this function is implemented using [consume], the function is _terminal_.
+ * In practice, this means that if adding new elements to the collection fails with an exception,
+ * that exception will be used for [cancelling][ReceiveChannel.cancel] the channel and rethrown.
+ *
+ * The intended use case for this function is collecting the remaining elements of a closed channel
+ * and processing them in a single batch.
+ *
+ * Example:
+ * ```
+ * val doContinue = AtomicBoolean(true)
+ *
+ * // Start the sender
+ * val channel = produce {
```

Comment:
Re using `produce` - could you confirm if our undestanding matches?

In theory we may want to avoid using `@ExperimentalCoroutinesApi` in samples, but the channels are the same level of lower-level abstractions, so if someone uses the channels, it's ok to make them aware of `produce`.